### ISSUE TITLE
Better error reporting for when memory accesses break the range check

### ISSUE THREAD

**Issue by hackaugusto** [0xPolygonMiden/miden-vm#833]
The following program works:

```
begin push.0 mem_load push.4294967295 mem_load end
```

the following fails:

```
begin push.0 mem_load push.4294967296 mem_load end
```

As explained to me by @bobbinth , this happens because at the end of the program execution, the gap between memory access must be at most `2**32`. The issue is that testing the above gives simply `not a 32-bit value`, and when `run`ing or `prove`ing with the release binary there is no reporting, which makes debugging this issue harder. The issue is to improve the error reporting so if someone encounters the issue, it is easy to tell what happened.

**Comment by bobbinth** [comment]
One way to fix this is to just set a hard limit on memory address - i.e., valid memory addresses can be in the range between `[0, 2^32)`. We can impost this first in the processor, and then later figure out if we can impose the same constraints via AIR.

**Comment by hackaugusto** [comment]
> valid memory addresses can be in the range between [0, 2^32).

If we do this, then we know that a memory pointer can not be a u64, and we can start using `u32` operations for everything. This would open some other optimizations, for example, the `mmr::unpack` could be implemented with the fixed `32` peaks https://github.com/0xPolygonMiden/miden-vm/pull/834/files#r1162724881 , since it would not be possible to represent the extra leaves in the VM.

**Comment by bobbinth** [comment]
> for example, the `mmr::unpack` could be implemented with the fixed `32` peaks https://github.com/0xPolygonMiden/miden-vm/pull/834/files#r1162724881 , since it would not be possible to represent the extra leaves in the VM.

I do think this will enable some optimizations, but I'm not sure the above case applies. Since absolute position is not a memory address, we can use a full field element to represent it, and since we describe the whole MMR with just its roots, we'll actually never have to use more than a few dozen memory addresses to describe an MMR.

**Comment by hackaugusto** [comment]
> Since absolute position is not a memory address, we can use a full field element to represent it

Makes sense, that basically means there will never be some sort of memory mapping going on, and values would be lazily loaded.

**Comment by Overcastan** [comment]
Closed by #1049
